src/torchcodec/decoders/_core/DeviceInterface.cpp

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "src/torchcodec/decoders/_core/DeviceInterface.h"
+#include <map>
+#include <mutex>
+
+namespace facebook::torchcodec {
+
+namespace {
+std::mutex g_interface_mutex;
+std::map<std::string, CreateDeviceInterfaceFn> g_interface_map;
+
+std::string getDeviceType(const std::string& device) {
+  size_t pos = device.find(':');
+  if (pos == std::string::npos) {
+    return device;
+  }
+  return device.substr(0, pos);
+}
+
+} // namespace
+
+bool registerDeviceInterface(
+    const std::string device_type,
+    CreateDeviceInterfaceFn create_interface) {
+  std::scoped_lock lock(g_interface_mutex);
+  TORCH_CHECK(
+      g_interface_map.find(device_type) == g_interface_map.end(),
+      "Device interface already registered for ",
+      device_type);
+  g_interface_map.insert({device_type, create_interface});
+  return true;
+}
+
+std::shared_ptr<DeviceInterface> createDeviceInterface(
```

Comment:
Here are in other places where we deal with this return value, we want to use `unique_ptr`. We want to enforce that only one entity owns the device at any one time. Unfortunately, this does mean we have to use ugly `std::move()` in a bunch of places.